README.MD
# uOSCORE / uEDHOC 

This repository contains C implementations  for constrained (and non-constrained) devices of the IETF protocols:

* [OSOCRE (RFC8613)](https://tools.ietf.org/html/rfc8613) and
* [EDHOC (draft-ietf-lake-edhoc-15 )](https://datatracker.ietf.org/doc/html/draft-ietf-lake-edhoc-15). 

Main features of uOSCORE and uEDHOC are their independence from the OS, cryptographic engine and in the case of uEDHOC transport protocol. Additionally,  uOSCORE and uEDHOC use only stack memory (no heap).

For more background and evaluation in terms of speed, RAM and flash requirements see our paper [The Cost of OSCORE and EDHOC for Constrained Devices](https://arxiv.org/pdf/2103.13832.pdf).

## How to Build and Link

* check the configurations in `makefile_config.mk` and adjust them if necessary
* run `make` 
* link the static library `build/libuoscore-uedhoc.a` in your project

## Project (Folder) Structure

```
.
|---cddl_models/
|---externals/
|---inc/
|---samples/
|---scripts/
|---src/
|---test/
|---test_vectors/
```
* The folder `cddl_models` contains CDDL models for all CBOR structures.
* The folder `externals` contains the external libraries and tools as git submodules.
* The folder `inc` contains all header file.
* The folder `samples` contains some usage examples.
* The folder `scripts` contains scripts for generatinc C code from CDDL models and converting the json formatted EDHOC test vectors to a C header
* The folder `src` contains all source file.
* The folder `test` contains automated tests. 
* The folder `test_vectors` contains tests vectors. 

## API and Usage Model

#### uOSCORE

The API of uOSCORE consists of three functions: 
* `oscore_context_init()`,
*  `coap2oscore()` and 
*  `oscore2coap()`.

`coap2oscore()` and `oscore2coap()` convert CoAP to OSCORE packets and vice versa. `oscore_context_init()` initializes the OSCORE security context. 

First, `oscore_context_init()` function needs to be called on the client and server side, then `coap2oscore()` and `oscore2coap()`  are called just before sending or receiving packets over the network.

<img src="oscore_usage.svg" alt="drawing" width="600"/>


#### uEDHOC

The API of uEDHOC consists of four functions:
*  `ephemeral_dh_key_gen()`
* `edhoc_initiator_run()`,
* `edhoc_responder_run()`,
* `edhoc_exporter()`,

`ephemeral_dh_key_gen()` is used to generate fresh ephemeral DH keys before running the protocol. This function requires a random seed suable for cryptographic purposes. `edhoc_initiator_run()` and `edhoc_responder_run() ` has to be called on the initiator and responder side respectively. They return the External Authorization data `EAD_x`,  the derived shared secret `PRK_out`.   `PRK_out` is used as input for `edhoc_exporter()` to derive application specific keys, e.g., OSCORE master secret and OSCORE master salt.

The EDHOC protocol requires the exchange of three messages (and an optional message 4) which is independent of the underlying message transport protocol. For example [appendix-A.2 in the EDHOC specification](https://datatracker.ietf.org/doc/html/draft-ietf-lake-edhoc-15#appendix-A.2) describes how  EDHOC can be transferred over CoAP, however CoAP is not mandatory. In order to be independent of the transport protocol uEDHOC uses two callback functions which need to be implemented by the user for handling the sending and receiving of messages. These functions are:

```c
/**
 * @brief   The user should call inside this function its send function. 
 *
 * 
 * @param   sock a pointer used to identify the rx chanel, 
 *          e.g. a socket handler
 * @param   data pointer to the data to be send
 * @param   data_len length of the data
 */
enum err tx(void *sock, uint8_t *data, uint32_t data_len);

/**
 * @brief   The user should call inside this function its receive 
 *          function and copy the received data in the buffer <data>.
 *          The length of the buffer <data> must be 
 *          checked before copying into it by using <data_len>. 
 *          After copying the length of the received data should be written 
 *          in <data_len>.
 *
 * 
 * @param   sock a pointer used to identify the rx chanel, 
 *          e.g. a socket handler
 * @param   data pointer to a buffer where the edhoc message must be copied 
 * @param   data_len length of the received data. When this function is 
 *          called inside EDHOC <data_len> is initialized with the actual 
 *          available length of the <data>.
 */
enum err rx(void *sock, uint8_t *data, uint32_t *data_len);
```



## Supported Cipher Suites

##### uOSCORE

| Algorithms                  |
| --------------------------- |
| AES-CCM-16-64-128,  SHA-256 |

##### uEDHOC


| Suit | Algorithms                                                                 |
| ---- | -------------------------------------------------------------------------- |
| 0    | AES-CCM-16-64-128, SHA-256, 8, X25519, EdDSA, AES-CCM-16-64-128, SHA-256   |
| 1    | AES-CCM-16-128-128, SHA-256, 16, X25519, EdDSA, AES-CCM-16-64-128, SHA-256 |
| 2    | AES-CCM-16-64-128, SHA-256, 8, P-256, ES256, AES-CCM-16-64-128, SHA-256    |
| 3    | AES-CCM-16-128-128, SHA-256, 16, P-256, ES256, AES-CCM-16-64-128, SHA-256  |



## Using the Samples

The samples contained in `samples/` are supposed to enable the test and verification of a real OSCORE or EDHOC message exchange between two entities. There are samples that can be run on microcontrollers, e.g. NRF5x (using ZephyrOS) and samples that can be run on Linux systems.

When using the samples for microcontroller , they have to be connected to an IPv6 network, which can be done with an IPv6 over BLE border router, using Bluetooth on the constrained devices. For further instructions on how to set up the border router on a Raspberry Pi, see [blerouter](externals/blerouter/README.MD).

* **OSCORE Between a Microcontroller (e.g. nrf52832) as a Server and Linux Host as a Client** 

  The `oscore_device` samples suggest a use of a microcontroller, (e.g. nrf52832) as an OSCORE server and a Linux machine as an OSCORE client. Therefore setting up the network as described in [blerouter](externals/blerouter/README.MD) is required.
  As this is a frequent error source: make sure that all addresses are set up correctly in the blerouter.sh script, the samples/oscore_linux/client/src/main.cpp file and the samples/oscore_device/server/prj.conf file.

  * Configure and build OSCORE server on a microcontroller (nrf52832) using the Zephyr's build tool west 

    * open a terminal and cd to samples/oscore/device/server 

    ```bash
    source <path to zephyr root>/zephyrproject/zephyr/zephyr-env.sh
    west build -b=nrf52dk_nrf52832
    west flash
    ```

    * The server is now reachable under the specified address and can be supplied with CoAP and OSCORE messages.

  * On a Raspberry Pi start externals/blerouter

    ```bash
    sudo ./blerouter.sh -d OscoreServer
    # if it works correctly you should see somthink like:
    # [18:27:47] scanning for devices
    # connect to D9:A4:16:41:CC:2D OscoreServer
    # [18:27:54] connected with D9:A4:16:41:CC:2D OscoreServer
    # [18:27:57] connected with D9:A4:16:41:CC:2D OscoreServer
    # ...
    ```

    

  * Configure OSCORE Linux client

    * open a terminal and cd to samples/oscore_linux/client

    * run ```make``` and ```./build/oscore_linux_client to send CoAP and OSCORE messages to the OSCORE server

* **OSCORE Between to Linux Hosts**

  Adjust the IP address of the other party in  samples/oscore_linux/client and samples/oscore_linux/server and run them.

* **EDHOC**

  Two parties are involved in a EDHOC message exchange -- an initiator (typically client) and a responder (typically server). The `edhoc_device` samples suggest the communication between a microcontroller and a Linux machine, with the network set up as described in [blerouter](externals/blerouter/README.MD). However, there can also be two Linux machines or two microcontrollers performing the EDHOC exchange, with the latter case requiring a second RPi set up as border router and connected to the second microcontroller. The addresses have to be changed accordingly in the prj.conf files. The following describes the setup for the EDHOC exchange between microcontroller and linux machine.

  * Microcontroller as initiator and Linux machine as responder

      * open a terminal and cd to samples/edhoc_linux/responder
      * make sure the IP addresses are correct
      * run ```make``` and ```./build/responder``` to build and run
      * the linux responder can now be supplied with an EDHOC requests
      * cd to samples/edhoc_device/initiator
      
      ```bash
      source <path to zephyr root>/zephyrproject/zephyr/zephyr-env.sh
      west build -b=nrf52dk_nrf52832
      west flash
      ```
      * set up the border router as described here [blerouter](externals/blerouter/README.MD)
      
      * the EDHOC message exchange should take place and the results should be printed on the linux machine - the message traffic can be captured and viewed using wireshark on the border router
      
  * Linux machine as initiator and Microcontroller as responder
    
    * open a terminal and cd to samples/edhoc_device/responder
    
    ```bash
    source <path to zephyr root>/zephyrproject/zephyr/zephyr-env.sh
    west build -b=nrf52dk_nrf52832
    west flash
    ```
    
    * set up the border router as described here [blerouter](externals/blerouter/README.MD)
    * the Microcontroller can now be supplied with an EDHOC request
    * cd to samples/edhoc_linux/initiator
    * run ```make``` and ```./build/initiator``` to build and run
    * the EDHOC message exchange should take place and the results should be printed on the linux machine - the message traffic can be captured and viewed using wireshark


## Using Different Cryptographic Libraries or Hardware Accelerators

The logic of uOSCORE and uEDHOC is independent form the cryptographic library, i.e., the cryptographic library can easily be exchanged by the user. For that the user needs to provide implementations for the functions specified in `crypto_wrapper.c`. 



change_log.md
#### v0.0.1 (12 Apr 2021)

#### v0.0.1 -> v0.1.0 (13 Apr 2021)
* update EDHOC to version v5 of the draft

#### v0.1.0 -> v1.0.0 (25 Jul 2022)
* update EDHOC to version v15 of the draft
* change the CBOR library to zcbor
* change the format of the text vectors
* Support for suites 0,1,2,3
* Support for methods 0,1,2,3
* use CI to run RFC tests
* Build uOSCORE and uEDHOC in a single static library
* Add support for X509 certificates
* improve API
* Add a sample showing the combined usage of OSCORE and EDHOC
* Add support for P256 and use mbedtls as crypto back-end
* Add replay window and sequence number checking for OSCORE
* add additional compiler warning flags 

inc/edhoc.h
/*
   Copyright (c) 2021 Fraunhofer AISEC. See the COPYRIGHT
   file at the top-level directory of this distribution.

   Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
   http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
   <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
   option. This file may not be copied, modified, or distributed
   except according to those terms.
*/

#ifndef EDHOC_H
#define EDHOC_H

#include <stdint.h>

#include "edhoc/edhoc_method_type.h"
#include "edhoc/messages.h"
#include "edhoc/suites.h"
#include "edhoc/hkdf_info.h"

#include "common/oscore_edhoc_error.h"
#include "common/byte_array.h"
#include "common/print_util.h"

/*define EDHOC_BUF_SIZES_RPK in order to use smaller buffers and save some RAM if need when RPKs are used*/
//#define EDHOC_BUF_SIZES_RPK
//#define EDHOC_BUF_SIZES_C509_CERT
#define EDHOC_BUF_SIZES_X509_CERT

#if defined EDHOC_BUF_SIZES_RPK
#define MSG_1_DEFAULT_SIZE 64
#define MSG_2_DEFAULT_SIZE 128
#define MSG_3_DEFAULT_SIZE 100
#define MSG_4_DEFAULT_SIZE 100
#define CIPHERTEXT2_DEFAULT_SIZE 100
#define CIPHERTEXT3_DEFAULT_SIZE 100
#define A_2M_DEFAULT_SIZE 200
#define M_3_DEFAULT_SIZE 200
#define CRED_DEFAULT_SIZE 128
#define SGN_OR_MAC_DEFAULT_SIZE 128
#define ID_CRED_DEFAULT_SIZE 20
#define PRK_3AE_DEFAULT_SIZE 100
#define CERT_DEFAUT_SIZE 128
#define CONTEXT_MAC_DEFAULT_SIZE 200
#define INFO_DEFAULT_SIZE 250
#define SIGNATURE_STRUCT_DEFAULT_SIZE 300
#endif

#if defined EDHOC_BUF_SIZES_C509_CERT
#define MSG_1_DEFAULT_SIZE 255
#define MSG_2_DEFAULT_SIZE 255
#define MSG_3_DEFAULT_SIZE 255
#define MSG_4_DEFAULT_SIZE 255
#define CIPHERTEXT2_DEFAULT_SIZE 255
#define CIPHERTEXT3_DEFAULT_SIZE 255
#define CIPHERTEXT4_DEFAULT_SIZE 255
#define A_2M_DEFAULT_SIZE 512
#define M_3_DEFAULT_SIZE 512
#define CRED_DEFAULT_SIZE 255
#define SGN_OR_MAC_DEFAULT_SIZE 128
#define ID_CRED_DEFAULT_SIZE 255
#define PLAINTEXT_DEFAULT_SIZE 255
#define CERT_DEFAUT_SIZE 255
#define CONTEXT_MAC_DEFAULT_SIZE 200
#define INFO_DEFAULT_SIZE 250
#define SIGNATURE_STRUCT_DEFAULT_SIZE 300
#endif

#if defined EDHOC_BUF_SIZES_X509_CERT
#define MSG_1_DEFAULT_SIZE 255
#define MSG_2_DEFAULT_SIZE 700
#define MSG_3_DEFAULT_SIZE 700
#define MSG_4_DEFAULT_SIZE 700
#define PLAINTEXT_DEFAULT_SIZE 1024
#define CIPHERTEXT2_DEFAULT_SIZE PLAINTEXT_DEFAULT_SIZE
#define CIPHERTEXT3_DEFAULT_SIZE PLAINTEXT_DEFAULT_SIZE
#define CIPHERTEXT4_DEFAULT_SIZE PLAINTEXT_DEFAULT_SIZE
#define A_2M_DEFAULT_SIZE 512
#define M_3_DEFAULT_SIZE 512
#define CRED_DEFAULT_SIZE 600
#define SGN_OR_MAC_DEFAULT_SIZE 128
#define ID_CRED_DEFAULT_SIZE 600
#define CERT_DEFAULT_SIZE 600
#define CONTEXT_MAC_DEFAULT_SIZE 1200
#define INFO_DEFAULT_SIZE 1200
#define SIGNATURE_STRUCT_DEFAULT_SIZE 1200
#endif

#define SUITES_MAX 5
#define ERR_MSG_DEFAULT_SIZE 64
#define P_256_PRIV_KEY_DEFAULT_SIZE 32
#define P_256_PUB_KEY_COMPRESSED_SIZE 33
#define P_256_PUB_KEY_UNCOMPRESSED_SIZE 65
#define P_256_PUB_KEY_X_CORD_SIZE 32
#define PK_DEFAULT_SIZE P_256_PUB_KEY_UNCOMPRESSED_SIZE
#define C_R_DEFAULT_SIZE 16
#define C_I_DEFAULT_SIZE 16
#define G_Y_DEFAULT_SIZE P_256_PUB_KEY_X_CORD_SIZE
#define G_X_DEFAULT_SIZE P_256_PUB_KEY_X_CORD_SIZE
#define G_R_DEFAULT_SIZE P_256_PUB_KEY_UNCOMPRESSED_SIZE
#define G_I_DEFAULT_SIZE P_256_PUB_KEY_UNCOMPRESSED_SIZE
#define DATA_2_DEFAULT_SIZE                                                    \
	(C_I_DEFAULT_SIZE + G_Y_DEFAULT_SIZE + C_R_DEFAULT_SIZE)
#define TH2_INPUT_DEFAULT_SIZE                                                 \
	(G_Y_DEFAULT_SIZE + C_R_DEFAULT_SIZE + HASH_DEFAULT_SIZE)
#define TH34_INPUT_DEFAULT_SIZE (HASH_DEFAULT_SIZE + PLAINTEXT_DEFAULT_SIZE)
#define ECDH_SECRET_DEFAULT_SIZE 32
#define DERIVED_SECRET_DEFAULT_SIZE 32
#define AD_DEFAULT_SIZE 256
#define PRK_DEFAULT_SIZE 32
#define ASSOCIATED_DATA_DEFAULT_SIZE 64
#define KID_DEFAULT_SIZE 8
#define HASH_DEFAULT_SIZE 32
#define AEAD_KEY_DEFAULT_SIZE 16
#define MAC_DEFAULT_SIZE 16
#define AEAD_IV_DEFAULT_SIZE 13
#define SIGNATURE_DEFAULT_SIZE 64
#define TH_ENC_DEFAULT_SIZE 42
#define ENCODING_OVERHEAD 6

struct other_party_cred {
	struct byte_array id_cred; /*ID_CRED_x of the other party*/
	struct byte_array cred; /*CBOR encoded credentials*/
	struct byte_array pk; /*authentication pub key of the other party */
	struct byte_array g; /*authentication static DH pub key of other party */
	struct byte_array ca; /*use only when authentication with certificates*/
	struct byte_array
		ca_pk; /*use only when authentication with certificates*/
};

struct edhoc_responder_context {
	bool msg4; /*if true massage 4 will be send by the responder*/
	struct byte_array c_r; /*connection identifier of the responder*/
	struct byte_array suites_r;
	struct byte_array g_y; /*ephemeral dh public key*/
	struct byte_array y; /*ephemeral dh secret key*/

	struct byte_array g_r; /* static DH pk -> use only with method 1 or 3*/
	struct byte_array r; /* static DH sk -> use only with method 1 or 3*/
	struct byte_array ead_2;
	struct byte_array ead_4;
	struct byte_array id_cred_r;
	struct byte_array cred_r;
	struct byte_array sk_r; /*sign key -use with method 0 and 2*/
	struct byte_array pk_r; /*coresp. pk to sk_r -use with method 0 and 2*/
	void *sock; /*pointer used as handler for sockets by tx/rx */
};

struct edhoc_initiator_context {
	bool msg4; /*if true massage 4 will be send by the responder*/
	struct byte_array c_i; /*connection identifier of the initiator*/
	enum method_type method;
	//uint8_t corr;
	struct byte_array suites_i;
	struct byte_array ead_1;
	struct byte_array ead_3;
	struct byte_array id_cred_i;
	struct byte_array cred_i;
	struct byte_array g_x; /*ephemeral dh public key*/
	struct byte_array x; /*ephemeral dh secret key*/
	struct byte_array g_i; /* static DH pk -> use only with method 2 or 3*/
	struct byte_array i; /* static DH sk -> use only with method 2 or 3*/
	struct byte_array sk_i; /*sign key use with method 0 and 2*/
	struct byte_array pk_i; /*coresp. pk to sk_r -use with method 0 and 2*/
	void *sock; /*pointer used as handler for sockets by tx/rx */
};

/**
 * @brief   Generates public and private ephemeral DH keys from a random seed. 
 *          
 *          IMPORTANT!!! PROVIDE A GOOD RANDOM SEED! 
 *
 * @param   curve DH curve to used
 * @param   seed a random seed
 * @param   sk pointer to a buffer where the secret key will be strored
 * @param   pk pointer to a buffer where the public key will be strored
 * @param   pk_size pointer to a variable with public key buffer size as input,
 *          and public key length as output.

 */
enum err __attribute__((weak))
ephemeral_dh_key_gen(enum ecdh_alg alg, uint32_t seed, uint8_t *sk, uint8_t *pk,
		     uint32_t *pk_size);

/**
 * @brief   Executes the EDHOC protocol on the initiator side
 * @param   c cointer to a structure containing initialization parameters
 * @param   cred_r_array containing elements of type other_party_cred used for
 *          the retrival of the other party (the responder) parameters at run
 *          time
 * @param   num_cred_r number of the elements in cred_r_array
 * @param   err_msg in case that an error message is received its contend is 
 *          provided to the caller though the err_msg
 * @param   ead_2 the received in msg2 additional data is provided to the 
 *          caller through ead_2
 * @param   ead_2_len length of ead_2
 * @param   prk_out the derived shared secret
 * @param   prk_out_len length of prk_out
 */
enum err edhoc_initiator_run(
	const struct edhoc_initiator_context *c,
	struct other_party_cred *cred_r_array, uint16_t num_cred_r,
	uint8_t *err_msg, uint32_t *err_msg_len, uint8_t *ead_2,
	uint32_t *ead_2_len, uint8_t *ead_4, uint32_t *ead_4_len,
	uint8_t *prk_out, uint32_t prk_out_len,
	enum err (*tx)(void *sock, uint8_t *data, uint32_t data_len),
	enum err (*rx)(void *sock, uint8_t *data, uint32_t *data_len));

/**
 * @brief   Executes the EDHOC protocol on the responder side
 * @param   c cointer to a structure containing initialization parameters
 * @param   cred_i_array containing elements of type other_party_cred used for 
 *          the retrival of the other party (the initiator) parameters at run 
 *          time
 * @param   num_cred_i number of the elements in cred_i_array
 * @param   err_msg in case that an error message is received its contend is 
 *          provided to the caller though the err_msg
 * @param   ead_1 the received in msg1 additional data is provided to the caller 
 *          through ead_1
 * @param   ead_1_len length of ead_1
 * @param   ead_3 the received in msg3 additional data is provided to the caller 
 *          through ead_3
 * @param   ead_3_len length of ead_3
 * @param   prk_out the derived shared secret
 * @param   prk_out_len length of prk_out
 */
enum err edhoc_responder_run(
	struct edhoc_responder_context *c,
	struct other_party_cred *cred_i_array, uint16_t num_cred_i,
	uint8_t *err_msg, uint32_t *err_msg_len, uint8_t *ead_1,
	uint32_t *ead_1_len, uint8_t *ead_3, uint32_t *ead_3_len,
	uint8_t *prk_out, uint32_t prk_out_len,
	uint8_t *client_pub_key, uint32_t *client_pub_key_size,
	uint8_t *c_i_bytes, uint32_t *c_i_bytes_len,
	enum err (*tx)(void *sock, uint8_t *data, uint32_t data_len),
	enum err (*rx)(void *sock, uint8_t *data, uint32_t *data_len));

/**
 * @brief Computes PRK_exporter from PRK_out
 * 
 * @param app_hash_alg 	the EDHOC hash algorithm
 * @param prk_out 		the product of a successful EDHOC execution
 * @param prk_out_len 	length of prk_out
 * @param prk_exporter 	pointer where the prk_exporter value will be written
 * @return enum err 0 or error code
 */
enum err prk_out2exporter(enum hash_alg app_hash_alg, uint8_t *prk_out,
			  uint32_t prk_out_len, uint8_t *prk_exporter);

/**
 * @brief Updates PRK_out
 * 
 * @param app_hash_alg 	the EDHOC hash algorithm
 * @param prk_out 		the product of a successful EDHOC execution 
 * @param prk_out_len 	length of prk_out
 * @param context		A context on which initiator and responder needs to 
 * 						agree in front
 * @param context_len	length of context
 * @param prk_out_new 	pointer where the prk_out_new value will be written
 * @return enum err 0 or error code
 */
enum err prk_out_update(enum hash_alg app_hash_alg, uint8_t *prk_out,
			uint32_t prk_out_len, uint8_t *context,
			uint32_t context_len, uint8_t *prk_out_new);

enum export_label {
	OSCORE_MASTER_SECRET = 0,
	OSCORE_MASTER_SALT = 1,
};

/**
 * @brief 	Computes key material to be used within the application, 
 * 			e.g., OSCORE master secret or OSCORE master salt
 * 
 * @param app_hash_alg		the application hash algorithm
 * @param label				an uint value defined by the application 
 * @param prk_exporter		PRK computed with prk_out2exporter()
 * @param prk_exporter_len 	length of prk_exporter
 * @param out 				the result of the computation,
 * 							e.g., OSCORE master secret or OSCORE master salt
 * @param out_len 			length of out
 * @return enum err 		0 or error code
 */
enum err edhoc_exporter(enum hash_alg app_hash_alg, enum export_label label,
			uint8_t *prk_exporter, uint32_t prk_exporter_len,
			uint8_t *out, uint32_t out_len);

#endif
